File: backend/src/routes/comments.js
import express from 'express';
const router = express.Router();
import { auth } from '../middleware/auth.js';
import Comment from '../models/Comment.js';
import Activity from '../models/Activity.js';

router.get('/task/:taskId', auth, async (req, res) => {
  try {
    const comments = await Comment.find({ taskId: req.params.taskId })
      .populate('userId', 'name email avatarUrl')
      .sort({ createdAt: 1 });

    res.json(comments.map(c => ({
      id: c._id.toString(),
      taskId: c.taskId.toString(),
      userId: c.userId._id.toString(),
      userName: c.userName,
      content: c.content,
      createdAt: c.createdAt.toISOString(),
      updatedAt: c.updatedAt ? c.updatedAt.toISOString() : undefined,
      parentId: c.parentId ? c.parentId.toString() : undefined,
    })));
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post('/', auth, async (req, res) => {
  try {
    const { taskId, content, parentId } = req.body;

    const comment = new Comment({
      taskId,
      userId: req.user._id,
      userName: req.user.name,
      content,
      parentId,
    });

    await comment.save();

    const activity = new Activity({
      userId: req.user._id,
      userName: req.user.name,
      action: 'commented on task',
      targetType: 'task',
      targetId: taskId,
    });
    await activity.save();

    res.status(201).json({
      id: comment._id.toString(),
      taskId: comment.taskId.toString(),
      userId: comment.userId.toString(),
      userName: comment.userName,
      content: comment.content,
      createdAt: comment.createdAt.toISOString(),
      parentId: parentId || undefined,
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.put('/:id', auth, async (req, res) => {
  try {
    const comment = await Comment.findById(req.params.id);

    if (!comment) {
      return res.status(404).json({ error: 'Comment not found' });
    }

    if (comment.userId.toString() !== req.user._id.toString()) {
      return res.status(403).json({ error: 'Not authorized' });
    }

    comment.content = req.body.content;
    comment.updatedAt = new Date();
    await comment.save();

    res.json({
      id: comment._id.toString(),
      content: comment.content,
      updatedAt: comment.updatedAt.toISOString(),
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.delete('/:id', auth, async (req, res) => {
  try {
    const comment = await Comment.findById(req.params.id);

    if (!comment) {
      return res.status(404).json({ error: 'Comment not found' });
    }

    if (comment.userId.toString() !== req.user._id.toString()) {
      return res.status(403).json({ error: 'Not authorized' });
    }

    await Comment.findByIdAndDelete(req.params.id);
    res.json({ success: true });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

File: backend/app/routes/comments.py
from fastapi import APIRouter, Depends, HTTPException, status
from app.auth import get_current_user, transform_record
from app.database import get_admin_client
from app.schemas.comment import CommentCreate, CommentUpdate

router = APIRouter()


@router.get("/task/{task_id}")
async def get_comments(task_id: str, current_user: dict = Depends(get_current_user)):
    supabase = get_admin_client()
    response = (
        supabase.table("comments")
        .select("*")
        .eq("task_id", task_id)
        .order("created_at", desc=False)
        .execute()
    )
    return [transform_record(c) for c in response.data]


@router.post("")
async def create_comment(
    body: CommentCreate, current_user: dict = Depends(get_current_user)
):
    supabase = get_admin_client()
    comment_data = {
        "task_id": body.taskId,
        "user_id": current_user["id"],
        "user_name": current_user["name"],
        "content": body.content,
        "parent_id": body.parentId,
    }

    response = supabase.table("comments").insert(comment_data).execute()
    comment = response.data[0]

    supabase.table("activities").insert(
        {
            "user_id": current_user["id"],
            "user_name": current_user["name"],
            "action": "commented on task",
            "target_type": "task",
            "target_id": body.taskId,
        }
    ).execute()

    return transform_record(comment)


@router.put("/{comment_id}")
async def update_comment(
    comment_id: str, body: CommentUpdate, current_user: dict = Depends(get_current_user)
):
    supabase = get_admin_client()
    response = supabase.table("comments").select("*").eq("id", comment_id).execute()
    if not response.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Comment not found"
        )

    comment = response.data[0]
    if comment["user_id"] != current_user["id"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized"
        )

    update_response = (
        supabase.table("comments")
        .update({"content": body.content})
        .eq("id", comment_id)
        .execute()
    )
    return transform_record(update_response.data[0])


@router.delete("/{comment_id}")
async def delete_comment(
    comment_id: str, current_user: dict = Depends(get_current_user)
):
    supabase = get_admin_client()
    response = (
        supabase.table("comments").select("user_id").eq("id", comment_id).execute()
    )
    if not response.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Comment not found"
        )

    comment = response.data[0]
    if comment["user_id"] != current_user["id"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized"
        )

    supabase.table("comments").delete().eq("id", comment_id).execute()
    return {"success": True}
